### PR TITLE
Sms user autocreate account 2707723

### PR DESCRIPTION
**Pull request by dpi**
https://www.drupal.org/node/2707723

Update register via SMS feature to use phone number provider

### CODE REVIEW

File: modules/sms_user/config/install/sms_user.settings.yml [hunk 1]
```diff
@@ -1,6 +1,15 @@
...
+    status: false
+    reply_message: ''
+  formatted:
+    status: false
+    incoming_messages: []
+    activation_email: true
+    reply:
+      status: false
+      message: ''
+      message_failure: ''
```
Comment (almaudoh) [inline modules/sms_user/config/install/sms_user.settings.yml]: These names are not very descriptive now that I see them in config.
`all_unknown_numbers` --> `register_unknown_numbers`
`formatted` --> `register_preformatted_message`
`activation_email` --> `send_activation_email`
`reply` --> `reply_message`

Comment (dpi) [inline modules/sms_user/config/install/sms_user.settings.yml]: Yeh not for review yet ;)

Comment (dpi) [inline modules/sms_user/config/install/sms_user.settings.yml]: The installed config you quoted has not been updated to match the schema. If you were to install the module at  that snapshot you would get invalid config. Schema is accurate.

It likely looks better when you're getting the variables, and looking at schema.

I am yet to find a better name for preformatted --- naming things is hard.

> all_unknown_numbers --> register_unknown_numbers
> formatted --> register_preformatted_message

register_ is implied by the parent; account_registration.

> register_preformatted_message

_message doesnt describe contents, as there is much more than message inside.

> reply --> reply_message

I mean we have reply.status and reply.message when using get. so this is redundant.

**Some examples of get calls using proper schema:**

```
account_registration.all_unknown_numbers.status
account_registration.all_unknown_numbers.reply.status
account_registration.all_unknown_numbers.reply.message
account_registration.formatted.status
account_registration.formatted.incoming_messages.SEQUENCE
account_registration.formatted.activation_email
account_registration.formatted.reply.status
account_registration.formatted.reply.message
```

**Comment by dpi** [comment]
tasks
- fix email requirement
- changed formatted terms to preformatted?
- remove 'all' prefix
- more tests

File: modules/sms_user/config/install/sms_user.settings.yml (hunk 1 above)
Comment (almaudoh) [inline modules/sms_user/config/install/sms_user.settings.yml]: `all_unknown_numbers` is not sufficiently explanatory. It still doesn't tell what it does. Same with `formatted` in the examples you have given. The lower rung names can be left as is but would only make meaning if the other two are self-explanatory.

File: modules/sms_user/src/AccountRegistration.php [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    // Password.
+    $password = user_password();
+    $user->setPassword($password);
+
+    $validate = $user->validate();
+    if ($validate->count() == 0) {
+      $user->save();
+
+      // @todo autoconfirm the number?
+      // @see https://www.drupal.org/node/2709911
```
Comment (almaudoh) [inline modules/sms_user/src/AccountRegistration.php:145]: Considering that this is the process for confirmation of a number, we can assume the number confirmed.

File: modules/sms_user/config/install/sms_user.settings.yml (hunk 1 above)
Comment (dpi) [inline modules/sms_user/config/install/sms_user.settings.yml]: Like I said... https://github.com/dpi/smsframework/pull/21#issuecomment-212504314 naming is TBD. + the config schema is missing label/descriptions.

I dont expect to overload the names with a full description of functionality, im leaning towards 'unknown_numbers'.

'formatted' needs a better name.

File: modules/sms_user/src/AccountRegistration.php (hunk 2 above)
Comment (dpi) [inline modules/sms_user/src/AccountRegistration.php:145]: My rationale is https://www.drupal.org/node/2709911. Discuss there

File: modules/sms_user/config/install/sms_user.settings.yml (hunk 1 above)
Comment (dpi) [inline modules/sms_user/config/install/sms_user.settings.yml]: Some ideas for naming:

**unrecognised** / unknown / new / unconfirmed / unbound / unassociated

**pattern match / pattern** / template (fill in the gaps)

**Comment by dpi** [comment]
New -> #22

**Pull request closed without merging**